File: 08/ahnjaewoo/1622.cpp
class Fancy {
private:
    vector<int> x;
		vector<tuple<int, string, int>> operations;
		// index, operation, value
		int curIdx;
		vector<tuple<int, int>> memo;
		// current calculated index, current value
public:
    int MAX = pow(10, 9) + 7;

    Fancy() {
        x.clear();
				curIdx = -1;
    }
    
    void append(int val) {
        x.push_back((val % MAX));
				memo.push_back({0, (val % MAX)});
				curIdx = x.size();
    }
    
    void addAll(int inc) {
				operations.push_back({curIdx, "add", inc});
    }
    
    void multAll(int m) {
				operations.push_back({curIdx, "mult", m});
    }
    
    int getIndex(int idx) {
        if (x.size() <= idx) return -1;
				int memoIdx = get<0>(memo[idx]);
				int result = get<1>(memo[idx]);
				for (int i = memoIdx; i < operations.size(); i++) {
					int targetIdx = get<0>(operations[i]);
					string operation = get<1>(operations[i]);
					int value = get<2>(operations[i]);
					if (idx < targetIdx) {
						if (operation == "add") {
							result = (int)(((unsigned long)(result % MAX) + (value % MAX)) % MAX);
						} else if (operation == "mult") {
							result = (int)(((unsigned long)(result % MAX) * (value % MAX)) % MAX);
						}
					}
				}
				memo[idx] = {operations.size(), result};
        return result;
    }
};

/**
 * Your Fancy object will be instantiated and called as such:
 * Fancy* obj = new Fancy();
 * obj->append(val);
 * obj->addAll(inc);
 * obj->multAll(m);
 * int param_4 = obj->getIndex(idx);
 */

File: 08/changjun/1622.py
class Fancy:
    MOD = 10**9 + 7

    def __init__(self):
        self.sequence = []
        self.add_accum = 0
        self.mul_accum = 1

    def append(self, val: int) -> None:
        normalized_val = (val - self.add_accum + self.MOD) * pow(self.mul_accum, self.MOD - 2, self.MOD) % self.MOD
        self.sequence.append(normalized_val)

    def addAll(self, inc: int) -> None:
        self.add_accum = (self.add_accum + inc) % self.MOD

    def multAll(self, m: int) -> None:
        self.add_accum = self.add_accum * m % self.MOD
        self.mul_accum = self.mul_accum * m % self.MOD

    def getIndex(self, idx: int) -> int:
        if idx >= len(self.sequence):
            return -1
        return (self.sequence[idx] * self.mul_accum + self.add_accum) % self.MOD
        


# Your Fancy object will be instantiated and called as such:
# obj = Fancy()
# obj.append(val)
# obj.addAll(inc)
# obj.multAll(m)
# param_4 = obj.getIndex(idx)